src/tools/miri/cargo-miri/src/phases.rs

```diff
@@ -501,6 +501,10 @@ pub fn phase_runner(mut binary_args: impl Iterator<Item = String>, phase: Runner
     // Set missing env vars. We prefer build-time env vars over run-time ones; see
     // <https://github.com/rust-lang/miri/issues/1661> for the kind of issue that fixes.
     for (name, val) in info.env {
+        // See <https://github.com/rust-lang/   /pull/   >.
```

Comment:
```suggestion
        // `CARGO_MAKEFLAGS` contains information about how to reach the
        // cargo-managed jobserver, but by the time the program is being run,
        // that jobserver no longer exists. Hence we shouldn't forward this.
        // (This does mean that Miri builds the final crate without a jobserver,
        // but that is still better than claiming that there is a jobserver at some
        // file descriptor when there actually isn't.)
```

Reply:
Thank you, this should be better.

> cargo-managed jobserver

I'd prefer *jobserver*, afaik it's not necessary managed by cargo.

Reply:
> Miri builds the final crate without a jobserver

Are there use cases then Miri should be invoked from external build system?

Reply:
So what about
```suggestion
        // `CARGO_MAKEFLAGS` contains information about how to reach the
        // jobserver, but when the program is being run, cargo does not provide
        // the jobserver file descriptors. Hence we shouldn't forward this information.
        // (This does mean that Miri builds the final crate without a jobserver,
        // but that is still better than claiming that there is a jobserver at some
        // file descriptor when there actually isn't.)
```

> Are there use cases then Miri should be invoked from external build system?

I have no idea.

Reply:
> cargo does not provide the jobserver file descriptors

I'm not sure, if it's true.

Reply:
If it wasn't true then why did you get errors in Miri with this PR? If that is not true then we have to go back to square 1 and figure out what is going on, because that's the only explanation I have so far.

Reply:
If Miri treated the same as rustc, cargo ([always](https://github.com/rust-lang/cargo/blob/94770a5586969bed3a50a0f5e36a7b7c5a122d0a/src/cargo/core/compiler/job_queue/mod.rs#L36-L38)) provides jobserver.

The problem - it's different at the time of compiling rustc and running. So using description of rust-build-time jobserver at running rustc is wrong. The same thing, probably, applies to Miri.

Reply:
Miri is not treated the same as rustc by cargo.

Miri is being set as the `target.runner`, so that cargo invokes Miri to *run* the program. rustc is not usually invoked there. It seems that by the time the `target.runner` is executed, the file descriptors for the jobserver don't exist any more (maybe cargo quit its jobserver?), or they don't get forwarded. At least that is the only explanation I have for why your check might fail.

Might be good to get a cargo dev to confirm. Cc @epage @weihanglo

Reply:
cargo-miri is an external subcommand, right?
https://github.com/rust-lang/cargo/pull/10511 looks relevant.
I seems like it's possible for external subcommands to inherit jobserver, maybe it's cargo-miri just not doing something correctly?

Reply:
Ah, wait, if cargo-miri runs as a part of `cargo run`, then this comment https://github.com/rust-lang/cargo/pull/10511#issuecomment-1092057057 looks more relevant, and this use case seems to not be supported yet.

Reply:
`cargo-miri` is an external subcommand then then invokes `cargo run` or `cargo test` again with a lot of environment variables set up such that cargo will call back to cargo-miri when invoking rustc or rustdoc or running any crate.

So yeah it sounds more like your 2nd link, where the jobserver is requested when cargo wants to run a crate.

Reply:
Let me open an issue as @weihanglo suggested.

Reply:
https://github.com/rust-lang/cargo/issues/12597